modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/stored/TransportPutStoredSearchTemplateAction.java

```diff
@@ -0,0 +1,116 @@
+import org.elasticsearch.cluster.AckedClusterStateUpdateTask;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.block.ClusterBlockException;
+import org.elasticsearch.cluster.block.ClusterBlockLevel;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.script.ScriptException;
+import org.elasticsearch.script.ScriptMetaData;
+import org.elasticsearch.script.ScriptService;
+import org.elasticsearch.script.StoredScriptSource;
+import org.elasticsearch.script.TemplateService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+public class TransportPutStoredSearchTemplateAction extends TransportMasterNodeAction<
+        PutStoredTemplateRequest, PutStoredTemplateResponse> {
+
+    private final TemplateService templateService;
+    private final int maxScriptSizeInBytes;
+
+    @Inject
+    public TransportPutStoredSearchTemplateAction(Settings settings,
+            TransportService transportService, ClusterService clusterService, ThreadPool threadPool,
+            ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver,
+            TemplateService templateService) {
+        super(settings, PutStoredTemplateAction.NAME, transportService, clusterService,
+                threadPool, actionFilters, indexNameExpressionResolver,
+                PutStoredTemplateRequest::new);
+        this.templateService = templateService;
+        maxScriptSizeInBytes = ScriptService.SCRIPT_MAX_SIZE_IN_BYTES.get(settings);
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected PutStoredTemplateResponse newResponse() {
+        return new PutStoredTemplateResponse();
+    }
+
+    @Override
+    protected void masterOperation(PutStoredTemplateRequest request, ClusterState state,
+            ActionListener<PutStoredTemplateResponse> listener) throws Exception {
+        if (request.content().length() > maxScriptSizeInBytes) {
+            throw new IllegalArgumentException("exceeded max allowed stored script size in bytes ["
+                    + maxScriptSizeInBytes + "] with size [" + request.content().length()
+                    + "] for script [" + request.id() + "]");
+        }
+
+        StoredScriptSource source = StoredScriptSource.parse(templateService.getTemplateLanguage(),
+                request.content(), request.xContentType());
+        try {
+            templateService.checkCompileBeforeStore(source);
+        } catch (IllegalArgumentException | ScriptException e) {
+            throw new IllegalArgumentException("failed to parse/compile stored search template ["
```

Comment:
In this case we're literally copy and pasting the code for storage. In this case I wanted to duplicate the code because I know the script code is going to drift from the template code significantly and I want us to keep the template code as is while we do it.